### PR TITLE
Separated steps and recreated test data

### LINKED ISSUE #29: Ensure the returned data tables have the correct data types
For example the `peak` column should be always an integer, etc.

The list of exact datatypes should be also contained in the galaxy tool wrapper.

### LINKED ISSUE #80: Replace `label` usage with `rt` to be consistent in variable naming.
The label is sometimes used for the retention time - which is actually which?

Seems like the label is actually the rt.

- [x] #138
- [x] #139
- [x] #118 
- [x] #140

_Originally posted by @hechth in https://github.com/RECETOX/recetox-aplcms/pull/78#discussion_r932157542_

### LINKED ISSUE #87: Simplify feature.align output
The `feature.align` function output 4 values: 

- two numbers `mz_tol_relative` and `rt_tol_relative` (which are fine),
- and two tables `aligned.ftrs` and `pk.times` with the following columns:

```
aligned.ftrs - average of m/z, average of rt, min of m/z, max of m/z, the sum of areas per sample
pk.times     - average of m/z, average of rt, min of m/z, max of m/z, the median of rt per sample
```

I think these two tables could be merged and output simplified. However, this is quite a major refactoring step across the whole package.

Further steps:
- [x] #108 
- [x] #123

### LINKED ISSUE #106: Extract function for rt correction on single sample given template
- [x] Extract function to compute the template
- [x] Extract function for single sample correction
- [x] Implement unit test for single sample correction

### LINKED ISSUE #108: Adapt recover.weaker to new feature.align output

### LINKED ISSUE #112: Ensure every feature table is a tibble
- [x] #113
- [x] #114
- [x] #115
- [x] #116
- [x] #87 
- [x] #136

### LINKED ISSUE #115: Refactor `profile.to.features` v2 -> Introduce and return a tibble

### LINKED ISSUE #116: Return a tibble from `recover.weaker`

### LINKED ISSUE #117: Adapt tests to new variables names
Re-create the expected outputs using the new function to fix the column names in files.

Should be done after #80

### LINKED ISSUE #123: Use tibbles in `create_aligned_feature_table` and subsequent function calls.

### LINKED ISSUE #129: Extract `compute_clusters` part from `adjust.time` and `feature.align`
This should be a standalone function on the same level as `adjust.time` so that it can be used also inside Galaxy as a standalone step.

- [ ] It has to be extracted from the two functions `adjust.time` and `feature.align` and introduced in the `unsupervised` and `hybrid methods`.
- [ ] The tests for `adjust.time` and `feature.align` have to be adjusted to take different input data and not perform also this step

### LINKED ISSUE #135: Use `get_num_workers` function in all test cases
yes - we can make an issue for it.

_Originally posted by @hechth in https://github.com/RECETOX/recetox-aplcms/pull/133#discussion_r951247446_

### LINKED ISSUE #140: Replace `pos` in `adjust.time` with `rt`

### PR DESCRIPTION
**Pull request by hechth**
Extracted `compute_clusters` from feature.align and adjust.time and built the new workflow in `unsupervised`.
Closes #129 

Updated tests for feature.align to test the new `create_aligned_feature_table` function.
Contains changes from @xtrojak which change the output format to be compatible with XCMS.
Closes #123 
Closes #108
Closes #117 

Includes the refactored time adjustment step.
Closes #106 
Closes #140

Replaced named lists with tibbles to use dataframes throughout the whole program.
Closes #135
Closes #116
Closes #115 
Closes #112
Closes #87 
Closes #80 
Closes #29 

This PR contains contributions from @maximskorik , @xtrojak , @wverastegui  and @zargham-ahmad .

### CODE REVIEW

File: R/adjust.time.R [hunk 1]
```diff
@@ -74,40 +74,70 @@ fill_missing_values <- function(orig.feature, this.feature) {
   return(this.feature)
 }
 
+compute_template <- function(extracted_features) {
+  num.ftrs <- sapply(extracted_features, nrow)
+  template <- which.max(num.ftrs)
+  message(paste("the template is sample", template))
```
Comment (xtrojak) [inline R/adjust.time.R:81]: Isn't this just a debug message?

File: R/adjust.time.R [hunk 2]
```diff
@@ -74,40 +74,70 @@ fill_missing_values <- function(orig.feature, this.feature) {
...
+
+  candi <- tibble::as_tibble(extracted_features[[template]]) |> dplyr::select(c(mz, rt))
+  template_features <- dplyr::bind_cols(candi, sample_id = rep(template, nrow(candi)))
+  return(tibble::as_tibble(template_features))
+}
+
+correct_time <- function(this.feature, template_features, mz_tol_relative, rt_tol_relative) {
```
Comment (xtrojak) [inline R/adjust.time.R:88]:
```suggestion
correct_time <- function(this_feature, template_features, mz_tol_relative, rt_tol_relative) {
```

Might be nice to unify variables naming conventions (underscore vs. dot), not just here but in general.

File: R/adjust.time.R [hunk 3]
```diff
@@ -74,40 +74,70 @@ fill_missing_values <- function(orig.feature, this.feature) {
...
+    orig.features <- this.feature
+    template <- unique(template_features$sample_id)[1]
+    j <- unique(this.feature$sample_id)[1]
+
+    if (j != template) {
+      this.comb <- compute_comb(template_features, this.feature)
+      sel <- compute_sel(this.comb, mz_tol_relative, rt_tol_relative)
+
+      if (length(sel) < 20) {
+        cat("too few, aborted")
```
Comment (xtrojak) [inline R/adjust.time.R]: Not very self-explanatory message. Also, I guess the program won't actually abort, right?

File: R/adjust.time.R [hunk 4]
```diff
@@ -119,59 +149,10 @@ adjust.time <- function(extracted_features,
...
-        extracted_features[[j]],
-        this.feature
-      )
-    }
-    this.feature
-  }
+  template_features <- compute_template(extracted_features)
 
+  corrected_features <- foreach::foreach(this.feature = extracted_features) %do% 
+  correct_time(this.feature, template_features, mz_tol_relative, rt_tol_relative)
```
Comment (xtrojak) [inline R/adjust.time.R]: I personally miss a visual indent here.

File: R/hybrid.R [hunk 5]
```diff
@@ -305,13 +312,20 @@ hybrid <- function(
     recover_min_count = recover_min_count
   )
 
-  message("**** second round time correction ****")
-  recovered_corrected <- adjust.time(
-    extracted_features = recovered$extracted_features,
+  message("**** third time computing clusters ****")
```
Comment (xtrojak) [inline R/hybrid.R:315]: Does there exist the second time computing clusters?

File: R/recover.weaker.R [hunk 6]
```diff
@@ -681,37 +683,36 @@ recover.weaker <- function(filename,
...
   )
 
   # IMPORTANT: THIS CODE SECTION COULD BE USED TO REPLACE COMPUTE_TARGET_TIMES FOR THE TEST CASES AND
   # IS A MASSIVE SIMPLIFICATION.
   # sp <- splines::interpSpline(
-  #   unique(extracted_features$pos) ~ unique(adjusted_features$rt),
+  #   unique(extracted_features$rt) ~ unique(adjusted_features$rt),
   #   na.action = na.omit
   # )
   # target_times <- predict(sp, aligned.ftrs[, "rt"])$y
```
Comment (xtrojak) [inline R/recover.weaker.R:711]: Do we want to keep this?

File: tests/testthat/test-adjust-time.R [hunk 7]
```diff
@@ -1,62 +1,60 @@
 patrick::with_parameters_test_that(
-  "adjust time test",
+  "get_template",
```
Comment (xtrojak) [inline tests/testthat/test-adjust-time.R]:
```suggestion
  "compute template",
```

File: tests/testthat/test-unsupervised.R [hunk 8]
```diff
@@ -1,23 +1,45 @@
...
+    actual <- result$recovered_feature_sample_table
+
+    # # This piece of code serves to re-introduce the actual filenames in the actual outputs.
+    # # This was needed for comparison in the previous test cases.
+    # levels(actual$sample) <- sapply(files, get_sample_name)
+    # actual <- actual |> dplyr::arrange_at(keys)
+    # expected <- expected |> dplyr::arrange_at(keys)
+
+    # actual_to_compare <- dplyr::select(actual, keys)
+    # expected_to_compare <- dplyr::select(expected, keys)
```
Comment (xtrojak) [inline tests/testthat/test-unsupervised.R]: Do we want to keep this?

**Review by xtrojak — APPROVED** [review]
Excellent job @hechth, thanks for putting things together. 
There are just several minor things, otherwise looks good!

**Review by maximskorik — APPROVED** [review]
Other than the questions raised by @xtrojak, looks good!

**Review by zargham-ahmad — APPROVED** [review]

File: R/adjust.time.R (hunk 1 above)
Comment (hechth) [reply R/adjust.time.R:81]: well it is also kind of interesting to know that I think but yes - primarily a debug message.

Could be changed to actually make the output more interesting or to harmonize what is actually written to the output.

File: R/recover.weaker.R (hunk 6 above)
Comment (hechth) [reply R/recover.weaker.R:711]: Yes - might be good to actually replace this part in the future and I feel like it doesn't really hurt if it is there.

File: tests/testthat/test-unsupervised.R (hunk 8 above)
Comment (hechth) [reply tests/testthat/test-unsupervised.R]: Nope

File: R/hybrid.R (hunk 5 above)
Comment (hechth) [reply R/hybrid.R:315]: will be fixed in hybrid rewrite

File: R/adjust.time.R (hunk 4 above)
Comment (hechth) [reply R/adjust.time.R]: Fixed in https://github.com/RECETOX/recetox-aplcms/pull/149/commits/2725a2bad62efa83557ada62b2b36b4ca95e2f99

File: R/adjust.time.R (hunk 3 above)
Comment (hechth) [reply R/adjust.time.R]: Fixed in https://github.com/RECETOX/recetox-aplcms/pull/149/commits/2725a2bad62efa83557ada62b2b36b4ca95e2f99

File: R/adjust.time.R (hunk 2 above)
Comment (hechth) [reply R/adjust.time.R:88]: Partially adressed in https://github.com/RECETOX/recetox-aplcms/pull/149/commits/2725a2bad62efa83557ada62b2b36b4ca95e2f99

**Pull request merged**